### PR TITLE
Design proposal for extractors

### PR DESCRIPTION
**Pull request by cmdoret**
This PR proposes an initial implementation of extractors as described in https://github.com/SDSC-ORD/gimie/issues/18#issuecomment-1380403967. The main changes are:
* The addition of an abstract class `gimie.sources.abstract.Extractor` which is used as the base for all extractors
* A working implementation for `gimie.sources.remote.GithubExtractor` which uses the Github API.
* A `gimie.project.Project` class which orchestrates the instantiation and serialization of extractors.

> Note: CI/CD tests currently fail due to rate limiting from the Github API, but succeed otherwise.

### CODE REVIEW

**Review by sabinem — APPROVED** [review]
The architecture has improved a lot in my opinion.

These are the main point, where I still see some possible improvements: 
- can we just have one extractor class per file in the `sources`, so that the code files in `sources` get only as long as they have to be?
- does it make sense to have a second python module next to `sources` so that we have fewer python files at the top level of `gimie`? What about having another module `graph` which includes `graph.py`, `namespaces.py`, `models.py` and `project.py`

These points are just suggestions. So to me it is also okay to merge it as it is now. Therefore I will already approve the PR.

Since this PR was about improving the overall design, I did not check every line of code but my review relates to the change of the architecture.

File: gimie/sources/remote.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    ) -> Union[Organization, Person]:
+        """Set the person or organization who owns the repository."""
+        if owner_type == "User":
+            return self.get_user(name)
+        elif owner_type == "Organization":
+            return self.get_organization(name)
+        else:
+            raise ValueError(f"Unknown Github owner type: {owner_type}.")
+
+    def get_user(self, name: str) -> Person:
```
[the remark this answers is not in the document]
Comment (cmdoret) [reply gimie/sources/remote.py]: Maes sense, I made those requests (`_get_{user,owner,contributors,organization}`) private

File: gimie/sources/remote.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        return [self.get_user(cont["login"]) for cont in conts]
+
+    def to_graph(self) -> Graph:
+        jd = GithubExtractorSchema().dumps(self)
+        g: Graph = Graph().parse(format="json-ld", data=str(jd))
+        g.bind("schema", SDO)
+        return g
+
+
+class GithubExtractorSchema(JsonLDSchema):
```
[the remark this answers is not in the document]
Comment (cmdoret) [reply gimie/sources/remote.py]: Agreed!

**Comment by cmdoret** [comment]
Thanks Sabine, excellent points ! I applied all suggested changes. In addition, when moving `graph.py` to the `graph/` module, I also renamed it to `operations`.

This was to avoid e.g. `from gimie.graph.graph import combine_graph`, which is now instead `from gimie.graph.operations import combine_graph`.

**Pull request merged**